Add testing and fix issues for weights_only load for LRScheduler

mikaylagawarecki opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is described in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.Counter`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
  Fixed the above in a BC-compatible manner by adding `OneCyclicLR._anneal_func_type` as a string attribute and removing `OneCyclicLR.anneal_func`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123775

mikaylagawarecki committed:
Add testing and fix issues for weights_only load for LRScheduler

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

As title, there were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
   The approach I took to fix this in this PR was to override the `state_dict` and `load_state_dict` methods to save the function name instead, as I think part of the ethos of `weights_only` is that we don't want the state_dict to contain functions. This might be BC-breaking if people rely on what is in the `state_dict`, however I don't think we document this anywhere, so I am hopeful that we could consider this a bugfix?
   
I wasn't sure how else to allowlist `__builtin__.getattr` (I assume this refers to the getattr of `OneCycleLR`, but I can't seem to allowlist this).
   



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
   The approach I took to fix this in this PR was to override the `state_dict` and `load_state_dict` methods to save the function name instead, as I think part of the ethos of `weights_only` is that we don't want the state_dict to contain functions. This might be BC-breaking if people rely on what is in the `state_dict`, however I don't think we document this anywhere, so I am hopeful that we could consider this a bugfix?
   
I wasn't sure how else to allowlist `__builtin__.getattr` (I assume this refers to the getattr of `OneCycleLR`, but I can't seem to allowlist this).
   



[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/lr_scheduler.py:
```diff
@@ -1761,6 +1761,18 @@ def _format_param(self, name, optimizer, param):
             return [param] * len(optimizer.param_groups)
 
+
+    def state_dict(self):
```
Use versioning here to ensure old `state_dict` can be loaded

albanD replied:
I think a better refactor is:
- Add a `self._anneal_func_type` that is the string from the user
- Add a `self._anneal_func` method on the class that will check the _anneal_func_type field and call the appropriate method or will call self.anneal_func when that exists (for BC)

That way you don't need to override state_dict/load_state_dict

mikaylagawarecki replied:
Made the suggested fix

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
   The approach I took to fix this in this PR was to override the `state_dict` and `load_state_dict` methods to save the function name instead, as I think part of the ethos of `weights_only` is that we don't want the state_dict to contain functions. This might be BC-breaking if people rely on what is in the `state_dict`, however I don't think we document this anywhere, so I am hopeful that we could consider this a bugfix?
   
I wasn't sure how else to allowlist `__builtin__.getattr` (I assume this refers to the getattr of `OneCycleLR`, but I can't seem to allowlist this).
   



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
   The approach I took to fix this in this PR was to override the `state_dict` and `load_state_dict` methods to save the function name instead, as I think part of the ethos of `weights_only` is that we don't want the state_dict to contain functions. This might be BC-breaking if people rely on what is in the `state_dict`, however I don't think we document this anywhere, so I am hopeful that we could consider this a bugfix?
   
I wasn't sure how else to allowlist `__builtin__.getattr` (I assume this refers to the getattr of `OneCycleLR`, but I can't seem to allowlist this).
   



[ghstack-poisoned]

albanD commented on test/optim/test_lrscheduler.py:
```diff
@@ -2285,6 +2286,41 @@ def test_lr_scheduler_verbose_deprecation_warning(self, LRClass):
             LRClass(self.opt)
 
 
+    @parametrize("LRClass", [
```
Can you make a one off test to make sure you can load an old state dict that is hardcoded? Just to make sure the BC handling is working as expected?

albanD reviewed (approved):
SGTM!
Just a small test to exercise the BC handling code

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
   The approach I took to fix this in this PR was to override the `state_dict` and `load_state_dict` methods to save the function name instead, as I think part of the ethos of `weights_only` is that we don't want the state_dict to contain functions. This might be BC-breaking if people rely on what is in the `state_dict`, however I don't think we document this anywhere, so I am hopeful that we could consider this a bugfix?
   
I wasn't sure how else to allowlist `__builtin__.getattr` (I assume this refers to the getattr of `OneCycleLR`, but I can't seem to allowlist this).
   



[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
   The approach I took to fix this in this PR was to override the `state_dict` and `load_state_dict` methods to save the function name instead, as I think part of the ethos of `weights_only` is that we don't want the state_dict to contain functions. This might be BC-breaking if people rely on what is in the `state_dict`, however I don't think we document this anywhere, so I am hopeful that we could consider this a bugfix?
   
I wasn't sure how else to allowlist `__builtin__.getattr` (I assume this refers to the getattr of `OneCycleLR`, but I can't seem to allowlist this).
   



[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
  Fixed the above in a BC-compatible manner by adding `OneCyclicLR._anneal_func_type` as a string attribute and removing `OneCyclicLR.anneal_func`




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add testing and fix issues for weights_only load for LRScheduler"

Fixes https://github.com/pytorch/pytorch/issues/98921

There were two issues detected:
- `MultiStepLR`: issue is descried in https://github.com/pytorch/pytorch/issues/98921, this is resolved by allowlisting `collections.OrderedDict`
- `OneCycleLR`: `state_dict['anneal_func']` is either `<function OneCycleLR._annealing_cos at 0x7f364186f5b0>` or 
`<function OneCycleLR._annealing_linear at 0x7f39aa483640>` depending on the `anneal_func` kwarg. 
   This leads to `WeightsUnpickler error: Unsupported class __builtin__.getattr` from the `weights_only` Unpickler.
   
  Fixed the above in a BC-compatible manner by adding `OneCyclicLR._anneal_func_type` as a string attribute and removing `OneCyclicLR.anneal_func`




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
